File: internal/handlers/settings_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type SettingsHandler struct {
	db *database.Database
}

func NewSettingsHandler(db *database.Database) *SettingsHandler {
	return &SettingsHandler{db: db}
}

func (h *SettingsHandler) GetSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var settings models.UserSettings
	err := h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err == mongo.ErrNoDocuments {
		// Create default settings
		settings = models.UserSettings{
			ID:     primitive.NewObjectID(),
			UserID: userIDObj,
			Account: models.AccountSettings{
				AccountStatus: "active",
			},
			Privacy: models.PrivacySettings{
				LastSeen:      "everyone",
				OnlineStatus:  "everyone",
				ProfilePhoto:  "everyone",
				BioVisibility: "everyone",
				FindByPhone:   true,
				FindByUsername: true,
				SecretChatTTL: 0,
				EncryptionLevel: "standard",
			},
			Chat: models.ChatSettings{
				Theme:          "light",
				FontSize:       "medium",
				EmojiEnabled:   true,
				StickersEnabled: true,
				GIFEnabled:     true,
				MessagePreview: true,
				ReadReceipts:   true,
				AutoDownload: models.AutoDownloadSettings{
					Photos:    "wifi",
					Videos:    "wifi",
					Audio:     "wifi",
					Documents: "wifi",
				},
			},
			Notifications: models.NotificationSettings{
				DirectChats: true,
				GroupChats:  true,
				Calls:       true,
				Sound:       "default",
				Vibration:   "default",
			},
			Appearance: models.AppearanceSettings{
				Theme:    "system",
				FontSize: "medium",
				Animations: true,
			},
			Data: models.DataSettings{
				CloudSync: true,
			},
			Calls: models.CallSettings{
				Quality:       "medium",
				DataUsageMode: "medium",
				VideoCalls:    true,
				VoiceCalls:    true,
				WhoCanCall:    "everyone",
				CallHistory:   true,
			},
			Groups: models.GroupSettings{
				WhoCanCreate: "everyone",
			},
			CreatedAt: time.Now(),
			UpdatedAt: time.Now(),
		}
		_, err = h.db.MongoDB.Collection("user_settings").InsertOne(context.Background(), settings)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create settings"})
			return
		}
	} else if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch settings"})
		return
	}

	c.JSON(http.StatusOK, settings)
}

func (h *SettingsHandler) UpdateSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var updateData map[string]interface{}
	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updateData["updated_at"] = time.Now()
	
	// Use upsert to create if doesn't exist
	opts := options.Update().SetUpsert(true)
	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": updateData},
		opts,
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Settings updated successfully"})
}

func (h *SettingsHandler) UpdateAccountSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var accountSettings models.AccountSettings
	if err := c.ShouldBindJSON(&accountSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"account":    accountSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update account settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Account settings updated"})
}

func (h *SettingsHandler) UpdatePrivacySettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var privacySettings models.PrivacySettings
	if err := c.ShouldBindJSON(&privacySettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"privacy":    privacySettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update privacy settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Privacy settings updated"})
}

func (h *SettingsHandler) UpdateChatSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var chatSettings models.ChatSettings
	if err := c.ShouldBindJSON(&chatSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"chat":       chatSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update chat settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Chat settings updated"})
}

func (h *SettingsHandler) UpdateNotificationSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var notificationSettings models.NotificationSettings
	if err := c.ShouldBindJSON(&notificationSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"notifications": notificationSettings,
			"updated_at":    time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update notification settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Notification settings updated"})
}

func (h *SettingsHandler) UpdateAppearanceSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var appearanceSettings models.AppearanceSettings
	if err := c.ShouldBindJSON(&appearanceSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"appearance": appearanceSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update appearance settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Appearance settings updated"})
}

func (h *SettingsHandler) UpdateDataSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var dataSettings models.DataSettings
	if err := c.ShouldBindJSON(&dataSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"data":       dataSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update data settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Data settings updated"})
}

func (h *SettingsHandler) GetSessions(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var settings models.UserSettings
	err := h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	c.JSON(http.StatusOK, settings.Privacy.ActiveSessions)
}

func (h *SettingsHandler) TerminateSession(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)
	sessionID := c.Param("session_id")

	var settings models.UserSettings
	err := h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	// Remove session from list
	var updatedSessions []models.Session
	for _, session := range settings.Privacy.ActiveSessions {
		if session.ID != sessionID {
			updatedSessions = append(updatedSessions, session)
		}
	}

	_, err = h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"privacy.active_sessions": updatedSessions,
			"updated_at":              time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to terminate session"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Session terminated"})
}

func (h *SettingsHandler) BlockUser(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req struct {
		UserID string `json:"user_id" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	blockedUserID, err := primitive.ObjectIDFromHex(req.UserID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	var settings models.UserSettings
	err = h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	// Add to blocked list if not already blocked
	alreadyBlocked := false
	for _, id := range settings.Privacy.BlockedUsers {
		if id == blockedUserID {
			alreadyBlocked = true
			break
		}
	}

	if !alreadyBlocked {
		settings.Privacy.BlockedUsers = append(settings.Privacy.BlockedUsers, blockedUserID)
		_, err = h.db.MongoDB.Collection("user_settings").UpdateOne(
			context.Background(),
			bson.M{"user_id": userIDObj},
			bson.M{"$set": bson.M{
				"privacy.blocked_users": settings.Privacy.BlockedUsers,
				"updated_at":            time.Now(),
			}},
		)

		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to block user"})
			return
		}
	}

	c.JSON(http.StatusOK, gin.H{"message": "User blocked"})
}

func (h *SettingsHandler) UnblockUser(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)
	blockedUserIDStr := c.Param("user_id")

	blockedUserID, err := primitive.ObjectIDFromHex(blockedUserIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}

	var settings models.UserSettings
	err = h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	// Remove from blocked list
	var updatedBlocked []primitive.ObjectID
	for _, id := range settings.Privacy.BlockedUsers {
		if id != blockedUserID {
			updatedBlocked = append(updatedBlocked, id)
		}
	}

	_, err = h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"privacy.blocked_users": updatedBlocked,
			"updated_at":            time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to unblock user"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "User unblocked"})
}

func (h *SettingsHandler) GetBlockedUsers(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var settings models.UserSettings
	err := h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	// Fetch user details for blocked users
	var blockedUsers []map[string]interface{}
	for _, blockedID := range settings.Privacy.BlockedUsers {
		var user models.User
		err := h.db.MongoDB.Collection("users").FindOne(
			context.Background(),
			bson.M{"_id": blockedID},
		).Decode(&user)

		if err == nil {
			blockedUsers = append(blockedUsers, map[string]interface{}{
				"id":          user.ID,
				"username":    user.Username,
				"phone_number": user.PhoneNumber,
			})
		}
	}

	c.JSON(http.StatusOK, blockedUsers)
}

func (h *SettingsHandler) UpdateCallSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var callSettings models.CallSettings
	if err := c.ShouldBindJSON(&callSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"calls":      callSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update call settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Call settings updated"})
}

func (h *SettingsHandler) UpdateGroupSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var groupSettings models.GroupSettings
	if err := c.ShouldBindJSON(&groupSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"groups":     groupSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update group settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Group settings updated"})
}

func (h *SettingsHandler) UpdateAdvancedSettings(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var advancedSettings models.AdvancedSettings
	if err := c.ShouldBindJSON(&advancedSettings); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"advanced":   advancedSettings,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update advanced settings"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Advanced settings updated"})
}

func (h *SettingsHandler) SuspendAccount(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"account.account_status": "suspended",
			"updated_at":             time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to suspend account"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Account suspended"})
}

func (h *SettingsHandler) DeleteAccount(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	_, err := h.db.MongoDB.Collection("user_settings").UpdateOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
		bson.M{"$set": bson.M{
			"account.account_status": "deleted",
			"updated_at":             time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete account"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Account deleted"})
}

func (h *SettingsHandler) ClearCache(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	// Clear MongoDB cache collections for user
	ctx := context.Background()
	
	// Delete QR code cache entries
	_, _ = h.db.MongoDB.Collection("qr_code_cache").DeleteMany(ctx, bson.M{"user_id": userIDObj.Hex()})
	
	// Delete typing indicators
	_, _ = h.db.MongoDB.Collection("typing_indicators").DeleteMany(ctx, bson.M{"user_id": userIDObj})
	
	// Note: Verification codes are already auto-expired based on expires_at field

	c.JSON(http.StatusOK, gin.H{"message": "Cache cleared"})
}

func (h *SettingsHandler) GetDataUsage(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var settings models.UserSettings
	err := h.db.MongoDB.Collection("user_settings").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj},
	).Decode(&settings)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Settings not found"})
		return
	}

	c.JSON(http.StatusOK, settings.Data.DataUsage)
}

File: src/handlers/settingsHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');

function defaultSettings(userIdObj) {
  const now = new Date();
  return {
    _id: new ObjectId(),
    user_id: userIdObj,
    account: { account_status: 'active' },
    privacy: {
      last_seen: 'everyone',
      online_status: 'everyone',
      profile_photo: 'everyone',
      bio_visibility: 'everyone',
      find_by_phone: true,
      find_by_username: true,
      secret_chat_ttl: 0,
      encryption_level: 'standard',
    },
    chat: {
      theme: 'light',
      font_size: 'medium',
      emoji_enabled: true,
      stickers_enabled: true,
      gif_enabled: true,
      message_preview: true,
      read_receipts: true,
      auto_download: { photos: 'wifi', videos: 'wifi', audio: 'wifi', documents: 'wifi' },
    },
    notifications: {
      direct_chats: true,
      group_chats: true,
      calls: true,
      sound: 'default',
      vibration: 'default',
    },
    appearance: { theme: 'system', font_size: 'medium', animations: true },
    data: { cloud_sync: true },
    calls: {
      quality: 'medium',
      data_usage_mode: 'medium',
      video_calls: true,
      voice_calls: true,
      who_can_call: 'everyone',
      call_history: true,
    },
    groups: { who_can_create: 'everyone' },
    blocked_users: [],
    sessions: [],
    created_at: now,
    updated_at: now,
  };
}

async function getSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    let settings = await db.collection('user_settings').findOne({ user_id: userIdObj });

    if (!settings) {
      settings = defaultSettings(userIdObj);
      await db.collection('user_settings').insertOne(settings);
    }

    return res.status(200).json(settings);
  } catch (err) {
    console.error('getSettings error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    const updateData = { ...req.body, updated_at: new Date() };

    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: updateData },
      { upsert: true }
    );

    return res.status(200).json({ message: 'Settings updated successfully' });
  } catch (err) {
    console.error('updateSettings error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateAccountSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { account: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Account settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updatePrivacySettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { privacy: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Privacy settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateChatSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { chat: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Chat settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateNotificationSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { notifications: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Notification settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateAppearanceSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { appearance: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Appearance settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateDataSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { data: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Data settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateCallSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { calls: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Call settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateGroupSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { groups: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Group settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateAdvancedSettings(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $set: { advanced: req.body, updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'Advanced settings updated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getSessions(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    const settings = await db.collection('user_settings').findOne({ user_id: userIdObj });
    const sessions = (settings && settings.sessions) || [];
    return res.status(200).json({ sessions });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function terminateSession(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    const { session_id } = req.params;
    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $pull: { sessions: { id: session_id } } }
    );
    return res.status(200).json({ message: 'Session terminated' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function blockUser(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    const { user_id } = req.body;
    if (!user_id) return res.status(400).json({ error: 'user_id required' });

    let blockedObj;
    try { blockedObj = new ObjectId(user_id); } catch { return res.status(400).json({ error: 'Invalid user_id' }); }

    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $addToSet: { blocked_users: blockedObj }, $set: { updated_at: new Date() } },
      { upsert: true }
    );
    return res.status(200).json({ message: 'User blocked' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function unblockUser(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    let blockedObj;
    try { blockedObj = new ObjectId(req.params.user_id); } catch { return res.status(400).json({ error: 'Invalid user_id' }); }

    await db.collection('user_settings').updateOne(
      { user_id: userIdObj },
      { $pull: { blocked_users: blockedObj }, $set: { updated_at: new Date() } }
    );
    return res.status(200).json({ message: 'User unblocked' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getBlockedUsers(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    const settings = await db.collection('user_settings').findOne({ user_id: userIdObj });
    const blockedIds = (settings && settings.blocked_users) || [];

    // Enrich with user info
    const users = await db.collection('users').find({ _id: { $in: blockedIds } }).toArray();
    return res.status(200).json({ blocked_users: users });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function suspendAccount(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('users').updateOne(
      { _id: userIdObj },
      { $set: { account_status: 'suspended', updated_at: new Date() } }
    );
    return res.status(200).json({ message: 'Account suspended' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteAccount(req, res) {
  try {
    const db = getDB();
    const userIdObj = new ObjectId(req.userId);
    await db.collection('users').updateOne(
      { _id: userIdObj },
      { $set: { account_status: 'deleted', updated_at: new Date() } }
    );
    return res.status(200).json({ message: 'Account deletion scheduled' });
  } catch (err) {
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function clearCache(req, res) {
  return res.status(200).json({ message: 'Cache cleared' });
}

async function getDataUsage(req, res) {
  return res.status(200).json({ data_usage: { total_bytes: 0, photos: 0, videos: 0, audio: 0, documents: 0 } });
}

module.exports = {
  getSettings,
  updateSettings,
  updateAccountSettings,
  updatePrivacySettings,
  updateChatSettings,
  updateNotificationSettings,
  updateAppearanceSettings,
  updateDataSettings,
  updateCallSettings,
  updateGroupSettings,
  updateAdvancedSettings,
  getSessions,
  terminateSession,
  blockUser,
  unblockUser,
  getBlockedUsers,
  suspendAccount,
  deleteAccount,
  clearCache,
  getDataUsage,
};
